### PR TITLE
[Impeller] grow glyph atlas instead of resizing when rect packer is full.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/138798

As far as I can tell, this gets us about as good as Skia, modulo the actual cost of rasterizing one of these glyphs is much more expensive for us.

The basic strategy is:

1. If the existing atlas texture has room (defined by the rect packer addRect call succeeding) append as many glyphs as we can to the old texture.
2. Check if there were any remaining glyphs. If not return.
3. Otherwise, double the size of the texture and create a new rect packer that represents only the increased area. So for example if you doubled a 100x100 texture to 100x200, then you'd create another 100x100 rect packer, assuming the old one is full.
4. Grow this size until all remaining glyphs fit. Store the new rect packer and the "height adjustment", which basically tells you where along the y-axis the rect packer offset starts.
5. Allocate this new texture, clear it, and then blit the contents of the old texture into the top left corner. Append all additional glyphs. The allows us to recycle all previously rendered glyphs and keep them in the texture, reducing the amount of software rasterization per frame.
6. Note; if the max atlas size would be exceeded, throw it away and start again.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I need to test the overflow behavior, I think I can do that just by creating giant glyphs. I'm also expecting many tests to fail, so lets see...

**Comment by cmkweber** [comment]
Couldn't it also expand horizontally when the max atlas size is reached (for instance 4096) and use a height-width adjustment? Then it could be 4096x4096.

**Comment by jonahwilliams** [comment]
definitely, we could keep expanding it in that direction. Initially I was trying to keep the same skyline rect packer so I wanted one dimension to stay constant for ease of use but I dropped that.

**Comment by jonahwilliams** [comment]
Sorry @chinmaygarde , you're it

**Comment by chinmaygarde** [comment]
> Allocate this new texture, clear it, and then blit the contents of the old texture into the top left corner.

Why clear the texture? The blit is going to achieve the same thing. And if not, then it indicates we are sampling from outside of the glyph region.

File: impeller/typographer/backends/skia/typographer_context_skia.cc [hunk 1]
```diff
@@ -376,141 +384,127 @@ std::shared_ptr<GlyphAtlas> TypographerContextSkia::CreateGlyphAtlas(
...
-  }
-  FML_DCHECK(!!blit_pass);
+  // attachments in most graphics drivers. For other textures, most framebuffer
+  // attachments have a much smaller size limit than the max texture size.
+  size_t byte_size =
+      new_texture->GetTextureDescriptor().GetByteSizeOfBaseMipLevel();
+  BufferView buffer_view =
+      host_buffer.Emplace(nullptr, byte_size, DefaultUniformAlignment());
+
+  ::memset(buffer_view.buffer->OnGetContents() + buffer_view.range.offset, 0,
```
Comment (chinmaygarde) [inline impeller/typographer/backends/skia/typographer_context_skia.cc:466]: Unclear about why this is necessary as the blit will clear the contents and we shouldn't be sampling from locations that aren't cleared.

Comment (chinmaygarde) [reply impeller/typographer/backends/skia/typographer_context_skia.cc:466]: Assuming this is necessary, let's add a trace around this?

File: impeller/typographer/backends/skia/typographer_context_skia.cc [hunk 2]
```diff
@@ -117,112 +120,111 @@ static SkImageInfo GetImageInfo(const GlyphAtlas& atlas, Size size) {
...
-    auto rect_packer = std::shared_ptr<RectanglePacker>(
-        RectanglePacker::Factory(current_size.width, current_size.height));
+    const std::vector<FontGlyphPair>& extra_pairs,
+    std::vector<Rect>& glyph_positions,
+    size_t glyph_index_start,
+    int64_t max_texture_height) {
+  // Because we can't grow the skyline packer horizontally, pick a reasonable
+  // large width for all atlases.
+  static constexpr auto kAtlasWidth = 4096u;
+  static constexpr auto kMinAtlasHeight = 1024;
```
Comment (chinmaygarde) [inline impeller/typographer/backends/skia/typographer_context_skia.cc:199]: Nit: Missing `u`. Especially since you are also using auto.

**Comment by jonahwilliams** [comment]
> Why clear the texture? The blit is going to achieve the same thing. And if not, then it indicates we are sampling from outside of the glyph region.

We do sample from outside the glyph region when the glyph is rotated/skewed. I tried clearing a 1px border around each glyph too but that still showed leaks. We have bugs in our glyph rendering logic :(

**Review by chinmaygarde — APPROVED** [review]
I didn't grasp point 5 of the strategy. But the implementation of the stated strategy lgtm.

**Comment by chinmaygarde** [comment]
> We have bugs in our glyph rendering logic :(

lol, fair.

**Comment by jonahwilliams** [comment]
Added more context to pt 5. What this does is allow us to preserve all previously rendered glyphs without using any additional software rasterization

File: impeller/typographer/backends/skia/typographer_context_skia.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/typographer/backends/skia/typographer_context_skia.cc:199]: Actually these should be signed since we pass them to ISize. Will remove auto.

File: impeller/typographer/backends/skia/typographer_context_skia.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/typographer/backends/skia/typographer_context_skia.cc]: Done
